zephyrproject-rtos/zephyr
# [RFC] Counter based system timer concept

This PR presents counter-based system timer. The whole idea is described in https://github.com/zephyrproject-rtos/zephyr/issues/12068.

Practical information:
- This PR contains changes form: https://github.com/zephyrproject-rtos/zephyr/pull/14794, https://github.com/zephyrproject-rtos/zephyr/pull/16469, https://github.com/zephyrproject-rtos/zephyr/pull/16591 as well as https://github.com/zephyrproject-rtos/zephyr/pull/16445. The actual implementation of the #12068 idea is in the last 2 commits.
- The code shows the concept. I have tried to make it as solid as possible, but please note that there is no optimization as well as some corner cases could be not handled correctly.

How to test it on your board:
1. Check if you have at least one counter device avaiable. If so, enable it.
2. Make sure, that your counter driver is initialized at PRE_KERNEL_1 level.
3. Enable CONFIG_COUNTER_TIMER and disable the old timer driver.
4. Add "zephyr,clock-source = <&ref-to-your-counter>" to choosen node of your board.
5. Compile and run the choosen test/sample. If you see assertions on startup, this means that the counter device you selected does not have all required features. In such case you have to enable CONFIG_COUNTER_GENERIC and instantiate it in the DTS. Use last commit from this PR as a reference.
